File: app/src/main/java/com/kickstarter/ui/viewholders/ProjectSearchResultViewHolder.java
package com.kickstarter.ui.viewholders;

import android.util.Pair;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.kickstarter.R;
import com.kickstarter.libs.KSString;
import com.kickstarter.libs.utils.ObjectUtils;
import com.kickstarter.libs.utils.ProjectUtils;
import com.kickstarter.models.Project;
import com.kickstarter.viewmodels.ProjectSearchResultHolderViewModel;
import com.squareup.picasso.Picasso;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import butterknife.Bind;
import butterknife.BindString;
import butterknife.ButterKnife;

import static com.kickstarter.libs.rx.transformers.Transformers.observeForUI;

public class ProjectSearchResultViewHolder extends KSViewHolder {
  protected final Delegate delegate;
  private final KSString ksString;
  private final ProjectSearchResultHolderViewModel.ViewModel viewModel;

  @Bind(R.id.search_result_deadline_countdown_text_view) TextView deadlineCountdownValueTextView;
  @Bind(R.id.search_result_deadline_unit_text_view) TextView deadlineCountdownUnitTextView;
  @Bind(R.id.project_name_text_view) TextView projectNameTextView;
  @Bind(R.id.project_image_view) ImageView projectImageView;
  @Bind(R.id.search_result_percent_funded_text_view) TextView percentFundedTextView;
  @Bind(R.id.search_result_funded_text_view) TextView fundedTextView;

  @BindString(R.string.discovery_baseball_card_stats_funded) String fundedString;
  @BindString(R.string.discovery_baseball_card_time_left_to_go) String toGoString;

  public interface Delegate {
    void projectSearchResultClick(ProjectSearchResultViewHolder viewHolder, Project project);
  }

  public ProjectSearchResultViewHolder(final @NonNull View view, final @NonNull Delegate delegate) {
    super(view);

    this.viewModel = new ProjectSearchResultHolderViewModel.ViewModel(environment());
    this.delegate = delegate;
    this.ksString = environment().ksString();

    ButterKnife.bind(this, view);

    this.viewModel.outputs.deadlineCountdownValueTextViewText()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(this.deadlineCountdownValueTextView::setText);

    this.viewModel.outputs.notifyDelegateOfResultClick()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(project -> this.delegate.projectSearchResultClick(this, project));

    this.viewModel.outputs.percentFundedTextViewText()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(this.percentFundedTextView::setText);

    this.viewModel.outputs.projectForDeadlineCountdownUnitTextView()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(p ->
        this.deadlineCountdownUnitTextView.setText(ProjectUtils.deadlineCountdownDetail(p, context(), this.ksString))
      );

    this.viewModel.outputs.projectNameTextViewText()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(this.projectNameTextView::setText);

    this.viewModel.outputs.projectPhotoUrl()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(this::setProjectImageUrl);

    this.fundedTextView.setText(this.fundedString);
  }

  @Override
  public void bindData(final @Nullable Object data) throws Exception {
    @SuppressWarnings("unchecked")
    final Pair<Project, Boolean> projectAndIsFeatured = ObjectUtils.requireNonNull((Pair<Project, Boolean>) data);
    this.viewModel.inputs.configureWith(projectAndIsFeatured);
  }

  private void setProjectImageUrl(final @NonNull String imageUrl) {
    Picasso.with(context()).load(imageUrl).into(this.projectImageView);
  }

  @Override
  public void onClick(final @NonNull View view) {
    this.viewModel.inputs.projectClicked();
  }
}

File: app/src/main/java/com/kickstarter/ui/viewholders/ProjectSearchResultViewHolder.kt
package com.kickstarter.ui.viewholders

import android.util.Pair
import android.view.View
import androidx.core.view.isInvisible
import androidx.core.view.isVisible
import com.kickstarter.R
import com.kickstarter.databinding.ProjectSearchResultViewBinding
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.libs.utils.extensions.deadlineCountdownDetail
import com.kickstarter.models.Project
import com.kickstarter.viewmodels.ProjectSearchResultHolderViewModel
import com.squareup.picasso.Picasso
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable

class ProjectSearchResultViewHolder(private val binding: ProjectSearchResultViewBinding, protected val delegate: Delegate) : KSViewHolder(binding.root) {
    private val ksString = requireNotNull(environment().ksString())
    private val viewModel = ProjectSearchResultHolderViewModel.ProjectSearchResultHolderViewModel(environment())
    private val disposables = CompositeDisposable()

    init {
        viewModel.outputs.deadlineCountdownValueTextViewText()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { binding.searchResultDeadlineCountdownTextView.text = it }
            .addToDisposable(disposables)

        viewModel.outputs.notifyDelegateOfResultClick()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { this.delegate.projectSearchResultClick(this, it) }
            .addToDisposable(disposables)

        viewModel.outputs.percentFundedTextViewText()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { binding.searchResultPercentFundedTextView.text = it }
            .addToDisposable(disposables)

        viewModel.outputs.projectForDeadlineCountdownUnitTextView()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { binding.searchResultDeadlineUnitTextView.text = it.deadlineCountdownDetail(context(), ksString) }
            .addToDisposable(disposables)

        viewModel.outputs.projectNameTextViewText()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { binding.projectNameTextView.text = it }
            .addToDisposable(disposables)

        viewModel.outputs.projectPhotoUrl()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { setProjectImageUrl(it) }
            .addToDisposable(disposables)

        binding.searchResultFundedTextView.setText(R.string.discovery_baseball_card_stats_funded)

        viewModel.outputs.displayPrelaunchProjectBadge()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe {
                binding.searchResultGroup.isInvisible = it
                binding.searchResultComingSoon.isVisible = it
            }.addToDisposable(disposables)
    }

    interface Delegate {
        fun projectSearchResultClick(viewHolder: KSViewHolder?, project: Project?)
    }

    override fun destroy() {
        viewModel.inputs.onCleared()
        disposables.clear()
        super.destroy()
    }

    @Throws(Exception::class)
    override fun bindData(data: Any?) {
        val projectAndIsFeatured = requireNotNull(data as Pair<Project, Boolean>?)
        viewModel.inputs.configureWith(projectAndIsFeatured)
    }

    private fun setProjectImageUrl(imageUrl: String) {
        Picasso.get().load(imageUrl).into(binding.projectImageView)
    }

    override fun onClick(view: View) {
        viewModel.inputs.projectClicked()
    }
}
